source/opt/folding_rules.cpp

```diff
@@ -180,8 +180,17 @@ std::vector<uint32_t> GetWordsFromNumericScalarOrVectorConstant(
 const analysis::Constant* ConvertWordsToNumericScalarOrVectorConstant(
     analysis::ConstantManager* const_mgr, const std::vector<uint32_t>& words,
     const analysis::Type* type) {
-  if (type->AsInteger() || type->AsFloat())
-    return const_mgr->GetConstant(type, words);
+  const spvtools::opt::analysis::Integer* int_type = type->AsInteger();
+
+  if (int_type && int_type->width() < 32 && !int_type->IsSigned()) {
```

Comment:
I recently added some utility functions you can use. See constant folding rules for SConvert.

Also you might need to handle 8bit integers as well. They are allowed in Spirv.

If we are casting from unsigned to signed, do we need to sign extend?

Reply:
> If we are casting from unsigned to signed, do we need to sign extend?

Ah yes we do, thanks for catching that.